Sources/SwiftWarningControl/WarningControlRegions.swift

```diff
@@ -0,0 +1,329 @@
+///
+/// The data structure represents these regions and their nesting relationships
+/// as an interval tree where interval nodes can only be nested or disjoint,
+/// and where a given interval corresponds to a diagnostic control region for one
+/// or more diagnostic group, with a behavior specifier for each.
+///
+/// Intervals cannot partially overlap, and each node's child intervals are always kept
+/// sorted. The tree has multiple top-level nodes (roots) representing file-level warning
+/// control regions.
+///
+/// Once the tree is computed, lookup of a diagnostic group behavior at a given position
+/// is performed by recursively descending into the child node containing
+/// the given position (located with a binary search of child nodes of a given parent node).
+/// Once the position's depth in the interval tree is reached, we walk back the
+/// traversal until we find the first containing region which specifies warning
+/// behavior control for the given diagnostic group id.
+///
+/// TODO: Capture global configuration from command-line arguments
+///       to represent global rules, such as `-Werror`, `-Wwarning`,
+///       and `-suppress-warnings` as *the* root region node.
+@_spi(ExperimentalLanguageFeatures)
+public struct WarningControlRegionTree {
+  /// Root region representing top-level (file) scope
+  private var rootRegionNode: WarningControlRegionNode
+
+  init(range: Range<AbsolutePosition>) {
+    rootRegionNode = WarningControlRegionNode(range: range)
+  }
+
+  /// Add a warning control region to the tree
+  mutating func addWarningGroupControl(
+    range: Range<AbsolutePosition>,
+    for diagnosticGroupIdentifier: DiagnosticGroupIdentifier,
+    behavior: WarningGroupBehavior
+  ) {
+    let newNode = WarningControlRegionNode(
+      range: range,
+      for: diagnosticGroupIdentifier,
+      behavior: behavior
+    )
+    insertIntoSubtree(newNode, parent: rootRegionNode)
+  }
+
+  /// Insert a region node into the appropriate position in a subtree
+  private func insertIntoSubtree(
+    _ node: WarningControlRegionNode,
+    parent: WarningControlRegionNode
+  ) {
+    // Check if the new region has the same boundaries as the parent
+    if parent.hasSameBoundaries(with: node) {
+      for (diagnosticGroupIdentifier, control) in node.warningGroupControls {
+        parent.addWarningGroupControl(for: diagnosticGroupIdentifier, behavior: control)
+      }
+      return
+    }
+
+    // Check if this should be a child of one of parent's children
+    if let containingChild = parent.findChildContaining(node.range.lowerBound) {
+      insertIntoSubtree(node, parent: containingChild)
+      return
+    }
+
+    // Check if this node should absorb any of the parent's children
+    var absorbedChildren: [WarningControlRegionNode] = []
+    for child in parent.children {
+      if node.contains(child) && !node.hasSameBoundaries(with: child) {
+        absorbedChildren.append(child)
+      }
+    }
```

Comment:
Can this ever happen? We are traversing the syntax tree top-down so nodes that are visited later should never contain any of the previously visited nodes.